File: tests/c/test_sequence.c
#include <assert.h>
#include <stdio.h>
#include "bitvector.h"

static void
test_concat_basic(void)
{
    BitVector *a = bv_new(5);
    BitVector *b = bv_new(3);

    bv_set(a, 0);
    bv_set(a, 2);
    bv_set(a, 4);

    bv_set(b, 1);

    BitVector *c = bv_concat(a, b);
    assert(c != NULL);
    assert(c->n_bits == 8);

    assert(bv_get(c, 0) == 1);
    assert(bv_get(c, 1) == 0);
    assert(bv_get(c, 2) == 1);
    assert(bv_get(c, 3) == 0);
    assert(bv_get(c, 4) == 1);

    assert(bv_get(c, 5) == 0);
    assert(bv_get(c, 6) == 1);
    assert(bv_get(c, 7) == 0);

    bv_free(a);
    bv_free(b);
    bv_free(c);
}

static void
test_concat_unaligned(void)
{
    BitVector *a = bv_new(70);
    BitVector *b = bv_new(70);

    bv_set(a, 0);
    bv_set(a, 69);

    bv_set(b, 1);
    bv_set(b, 68);

    BitVector *c = bv_concat(a, b);
    assert(c != NULL);
    assert(c->n_bits == 140);

    assert(bv_get(c, 0) == 1);
    assert(bv_get(c, 69) == 1);
    assert(bv_get(c, 70 + 1) == 1);
    assert(bv_get(c, 70 + 68) == 1);

    bv_free(a);
    bv_free(b);
    bv_free(c);
}

static void
test_repeat_basic(void)
{
    BitVector *a = bv_new(4);
    bv_set(a, 0);
    bv_set(a, 3);

    BitVector *r = bv_repeat(a, 3);
    assert(r != NULL);
    assert(r->n_bits == 12);

    int expected[] = {1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1};

    for (int i = 0; i < 12; i++) {
        assert(bv_get(r, i) == expected[i]);
    }

    bv_free(a);
    bv_free(r);
}

static void
test_repeat_zero(void)
{
    BitVector *a = bv_new(10);
    BitVector *r = bv_repeat(a, 0);

    assert(r != NULL);
    assert(r->n_bits == 0);

    bv_free(a);
    bv_free(r);
}

int
main(void)
{
    setvbuf(stdout, NULL, _IONBF, 0);

    test_concat_basic();
    printf("test_concat_basic: OK\n");

    test_concat_unaligned();
    printf("test_concat_unaligned: OK\n");

    test_repeat_basic();
    printf("test_repeat_basic: OK\n");

    test_repeat_zero();
    printf("test_repeat_zero: OK\n");

    return 0;
}

File: tests/python/test_sequence.py
import unittest
from cbits import BitVector

class TestSequenceOps(unittest.TestCase):
    def test_concat_basic(self):
        a = BitVector(5)
        b = BitVector(3)

        a.set(0); a.set(2); a.set(4)
        b.set(1)

        c = a + b
        self.assertIsInstance(c, BitVector)
        self.assertEqual(len(c), 8)

        expected = [1,0,1,0,1, 0,1,0]
        self.assertEqual(expected, [c.get(i) for i in range(8)])

    def test_concat_unaligned(self):
        a = BitVector(70)
        b = BitVector(70)

        a.set(0); a.set(69)
        b.set(1); b.set(68)

        c = a + b
        self.assertEqual(len(c), 140)
        self.assertTrue(c.get(0))
        self.assertTrue(c.get(69))
        self.assertTrue(c.get(70 + 1))
        self.assertTrue(c.get(70 + 68))

    def test_concat_type_error(self):
        a = BitVector(5)
        with self.assertRaises(TypeError):
            _ = a + 123

    def test_repeat_basic(self):
        a = BitVector(4)
        a.set(0); a.set(3)

        r = a * 3
        self.assertEqual(len(r), 12)

        expected = [1,0,0,1] * 3
        self.assertEqual(expected, [r.get(i) for i in range(12)])

    def test_repeat_zero(self):
        a = BitVector(10)
        r = a * 0
        self.assertEqual(len(r), 0)

    def test_repeat_negative(self):
        a = BitVector(5)
        with self.assertRaises(ValueError):
            _ = a * -1

    def test_repeat_left_mul(self):
        a = BitVector(3)
        a.set(1)

        r = 4 * a
        self.assertEqual(len(r), 12)
        expected = [0,1,0] * 4
        self.assertEqual(expected, [r.get(i) for i in range(12)])

if __name__ == "__main__":
    unittest.main()
